rust-runtime/aws-smithy-checksums/src/body.rs

```diff
@@ -0,0 +1,444 @@
+    pub fn trailer_length(&self) -> u64 {
+        let trailer_name_size_in_bytes = self.checksum.header_name().as_str().len() as u64;
+        let base64_encoded_checksum_size_in_bytes = base64::encoded_length(self.checksum.size());
+
+        trailer_name_size_in_bytes
+            // HTTP trailer names and values may be separated by either a single colon or a single
+            // colon and a whitespace. In the AWS Rust SDK, we use a single colon.
+            + ":".len() as u64
+            + base64_encoded_checksum_size_in_bytes
+    }
+
+    fn poll_inner(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Bytes, aws_smithy_http::body::Error>>> {
+        let this = self.project();
+        let inner = this.inner;
+        let mut checksum = this.checksum;
+
+        match inner.poll_data(cx) {
+            Poll::Ready(Some(Ok(data))) => {
+                if let Err(e) = checksum.update(&data) {
+                    return Poll::Ready(Some(Err(e)));
+                }
+
+                Poll::Ready(Some(Ok(data)))
+            }
+            Poll::Ready(None) => Poll::Ready(None),
+            Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e))),
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+impl http_body::Body for ChecksumBody<SdkBody> {
+    type Data = Bytes;
+    type Error = aws_smithy_http::body::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        self.poll_inner(cx)
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap<HeaderValue>>, Self::Error>> {
+        let this = self.project();
+        match (
+            this.checksum.headers(),
+            http_body::Body::poll_trailers(this.inner, cx),
+        ) {
+            // If everything is ready, return trailers, merging them if we have more than one map
+            (Ok(outer_trailers), Poll::Ready(Ok(inner_trailers))) => {
+                let trailers = match (outer_trailers, inner_trailers) {
+                    // Values from the inner trailer map take precedent over values from the outer map
+                    (Some(outer), Some(inner)) => Some(append_merge_header_maps(inner, outer)),
+                    // If only one or neither produced trailers, just combine the `Option`s with `or`
+                    (outer, inner) => outer.or(inner),
+                };
+                Poll::Ready(Ok(trailers))
+            }
+            // If the inner poll is Ok but the outer body's checksum callback encountered an error,
+            // return the error
+            (Err(e), Poll::Ready(Ok(_))) => Poll::Ready(Err(e)),
+            // Otherwise return the result of the inner poll.
+            // It may be pending or it may be ready with an error.
+            (_, inner_poll) => inner_poll,
+        }
+    }
```

Comment:
this is really cool! (how checksum body uses structured trailers _then_ AWS chunked writes the trailers into the body)